Update ERC725 Page (#456)
lukso-network/docs: docs/standards/lsp-background/erc725.md

## Introduction

> EOAs (Externally Owned Accounts) are controlled by private keys and can perform specific operations such as **CALL** to interact with other addresses and **CREATE** to create other smart contracts. However, they do not have any internal logic or storage capabilities.

Reviewer: maybe you could put a link on the words "call" and "create"? To this pages or smtgh similar:

- https://ethereum.org/en/developers/docs/evm/opcodes/
- https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a9-create-operations

I'm not sure if we did it somewhere else already

> Smart contracts, on the other hand, can be controlled by addresses or other smart contracts, which provide a flexible ownership management mechanism. They can store data and perform more complex operations, making them useful for a wide range of use cases, including digital assets, decentralized applications, DAO and smart contract based accounts.

Reviewer: i suggest to replace "controlled by addresses" to "controlled by EOAs" because in the paragraph above, you explain what an EOA is, not what an "address" is. So reader will be like: "Ok, EOA, I know what it is. But what is an address?" Bc it is not explained above that EOA ~~ addresses

## ERC725X

> There is a growing need for more operations beyond the basic **CALL** and **CREATE** functionality of an externally owned account (EOA). These two operations allows to interact with other addresses on the blockchain and perform tasks such as calling functions, transferring native tokens, and creating new contracts. As the number of use cases for smart contracts increases, so too does the need for additional functionality.

Reviewer 2: ```suggestion
There is a growing need for more operations beyond the basic **CALL** and **CREATE** functionality of an externally owned account (EOA). These two operations enables interaction with other addresses on the blockchain to perform tasks such as calling functions, transferring native tokens, and creating new contracts. As the number of use cases for smart contracts increases, so too does the need for additional functionality.
```

> There is a growing need for more operations beyond the basic **CALL** and **CREATE** functionality of an externally owned account (EOA). These two operations enables interaction with other addresses on the blockchain to perform tasks such as calling functions, transferring native tokens, and creating new contracts. As the number of use cases for smart contracts increases, so too does the need for additional functionality.

Reviewer 3: ```suggestion
There is a growing need for more functionalities beyond the basic **CALL** and **CREATE** that externally owned accounts (EOA) can do. These two opcodes enables interaction with other addresses on the blockchain to perform tasks such as calling functions, transferring native tokens, and creating new contracts. As the number of use cases for smart contracts increases, so does the need for additional functionality.
```

> One such operation is the ability to manipulate storage of a contract and execute other contracts logic in the caller context, which is not possible with EOA. Additionally, the ability to create contracts at predetermined addresses, as well as, performing calls without causing any state changes are also important features that are currently not available to EOAs.

Reviewer 2: I wonder if discussing caller context in this example is going a bit too technical. Could just start super basic - EOAs don't have any data store which limits the functionality they can perform

Author: It's the motivation for `ERC725X`, but yeah ill try to simplify

Reviewer 2: Ah okay well even going into more detail about what that means could be cool

|        2         |  [`CREATE2`](https://eips.ethereum.org/EIPS/eip-1014)  | Create a new smart contract based on the contract address, bytecode and the salt. The address can be predetermined. |
|        3         | [`STATICCALL`](https://eips.ethereum.org/EIPS/eip-214) | Calls another smart contract while disallowing any modification to the state during the call.                       |
|        4         | [`DELEGATECALL`](https://eips.ethereum.org/EIPS/eip-7) | Runs the function from another contract, but use the context of the current contract.                               |

## ERC725Y Generic Data Key/Value Store

Using smart contracts to store data on the blockchain can be beneficial when that data will be used or referenced by other smart contracts.

One limitation of this approach is that once a smart contract is deployed with a specific set of variables containing data, it cannot be modified to include new variables. This can be a problem if additional data needs to be stored in the future, as it would require deploying a new version of the contract.

> For example, if a token contract is deployed with variables for the token name, symbol, and decimals, it would not be possible to add new variables for the creators or future collaborators.

Reviewer 3: ```suggestion
For example, if a token contract is deployed with variables for the token name, symbol, and decimals, it is not possible to add new variables for the creators or future collaborators.
```

What is needed is a dynamic way to attach data to a contract even after it has been deployed.

### Specification

> ERC725Y solves this problem by standardizing a mapping of **data keys** to **data values** to store data dynamically, and to have the ability to add or remove data across time without the need of redeploying the contract.

Reviewer 3: ```suggestion
ERC725Y solves this problem by standardizing a mapping of **data keys** to **data values** to store data dynamically and to have the ability to add or remove data across time without the need to redeploy the contract.
```

> Instead of representing the variable with a specific name such as `age`, or `name`, etc .. They are referenceed with `bytes32`. Same as solidity treat variables under the hood.

Reviewer 3: Not sure if we can states "data key" at this point?

```suggestion
Instead of representing the variable with a specific name such as `age`, or `name`, etc .. They are referenced as `bytes32` data key. Same as solidity treat variables under the hood.
```

If you feel not, you can discard this comment :)

:::note

See the section **["Layout of State Variables in Storage"](https://docs.soliditylang.org/en/v0.8.11/internals/layout_in_storage.html)** in the Solidity documentation for more information about the structure of smart contract storage.

:::

> #### Normal contract Vs ERC725Y

Reviewer 3: ```suggestion
#### Normal contract vs ERC725Y
```

:::

This sub-standard enables the account to hold arbitrary data through a generic data key/value store. It gives flexibility to the contract storage. Developers can access the data stored in the contract via data keys instead of referencing the storage slot where the data resides.
#### Normal contract Vs ERC725Y

![ERC725Y Generic Data Key/Value Store](../../../static/img/standards/erc725/erc725y.jpeg)
![Smart contract vs ERC725Y](../../../static/img/standards/erc725/SmartContractVsERC725Y.jpeg)

As defined below, the data keys are represented as **bytes32** and the data values are represented as **bytes**. This allows more flexibility and support for complex variable types.
  
Difference between Standard Smart Contract Storage and ERC725Y - Generic Data Key/Value Storage:
Thanks to ERC725Y, contracts become more interoperable, as their storage is represented in the same way. Contracts and interfaces can then read and write data from or to the storage in the same manner via the functions `getData(...)` and `setData(...)`.

> ![ERC725Y Generic Data Key/Value Store](../../../static/img/standards/erc725/erc725y-difference.jpeg)

Reviewer: it looks like you don't use this image anymore - can you also delete this image from the repo?

> Since the data key/value pair is set in the storage as **bytes32**/**bytes**, a representation of this data should be standardized to avoid having each user writing the data in a different way and to have the ability to getData from other people contract storage in a unified way.

Reviewer 3: ```suggestion
Since the data is set in the contract's storage as **bytes32**/**bytes** data key/value pair, a representation of this data should be standardized to avoid having each user writing the data in a different way and to have the ability to getData from other people's contract storage in a unified way.
```
